### PR TITLE
Add Fine-tuning a Multimodal Model Using SFT (Single or Multi-Image Dataset) guide to docs

### PR DESCRIPTION
**Pull request by sergiopaniego**
# What does this PR do?

This PR adds a guide for Fine-tuning a Multimodal Model Using SFT (Single or Multi-Image Dataset) that is built from [the sft_vlm_gemma3.py](https://github.com/huggingface/trl/blob/main/examples/scripts/sft_vlm_gemma3.py) example script.

TODO:

- [ ] Add example results.
- [ ] Move diagrams to a persistent option.

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/trl/tree/main/docs).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: docs/source/training_vlm_sft.md [hunk 1]
```diff
@@ -0,0 +1,386 @@
+# Fine-tuning a Multimodal Model Using SFT (Single or Multi-Image Dataset)
+
+![Single Image + Text](./training_vlm_sft_training_procedure_single_image.png)
```
Comment (qgallouedec) [inline docs/source/training_vlm_sft.md]: could you please use https://huggingface.co/datasets/trl-lib/documentation-images for the images? I've just sent you an invite to join the org

File: docs/source/training_vlm_sft.md [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+
+This guide serves as a **detailed walkthrough** and complements the existing [VLM SFT script](https://github.com/huggingface/trl/blob/main/examples/scripts/sft_vlm_gemma3.py). If you're already familiar with the concepts, you can use the script directly.  
+
+We demonstrate the fine-tuning process using two datasets, but these principles extend to other **Vision-Language Models (VLMs)** and datasets.  
+
+## Understanding the Datasets  
+
+To address both **Single Image + Text** and **Multi-Image + Text** scenarios, we use two datasets that are well-suited for this task.  
+
+### **HuggingFaceH4/llava-instruct-mix-vsft Dataset (Image + Text)**
```
Comment (qgallouedec) [inline docs/source/training_vlm_sft.md:20]:
```suggestion
### HuggingFaceH4/llava-instruct-mix-vsft Dataset (Image + Text)
```

File: docs/source/training_vlm_sft.md [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+The model (referred to as the **"assistant"**) responds based on both the **visual and textual information** shared by the user. This dataset is particularly useful for training multimodal models to **understand and generate responses based on images and text**.  
+
+<iframe
+  src="https://huggingface.co/datasets/HuggingFaceH4/llava-instruct-mix-vsft/embed/viewer/default/train"
+  frameborder="0"
+  width="100%"
+  height="560px"
+></iframe>
+
+### **FanqingM/MMIU-Benchmark Dataset (Multi-Image + Text)**
```
Comment (qgallouedec) [inline docs/source/training_vlm_sft.md:33]:
```suggestion
### FanqingM/MMIU-Benchmark Dataset (Multi-Image + Text)
```

File: docs/source/training_vlm_sft.md [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+This dataset is designed for tasks where the model must reason over multiple images to generate an informed response based on both visual and textual inputs.
+
+<iframe
+  src="https://huggingface.co/datasets/FanqingM/MMIU-Benchmark/embed/viewer/default/test"
+  frameborder="0"
+  width="100%"
+  height="560px"
+></iframe>
+
+## **Developing a Fine-Tuning Script for Multimodal SFT**
```
Comment (qgallouedec) [inline docs/source/training_vlm_sft.md:51]:
```suggestion
## Developing a Fine-Tuning Script for Multimodal SFT
```

File: docs/source/training_vlm_sft.md [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+  frameborder="0"
+  width="100%"
+  height="560px"
+></iframe>
+
+## **Developing a Fine-Tuning Script for Multimodal SFT**  
+
+In this section, we build the script needed to fine-tune a multimodal model for both **Single Image + Text** and **Multi-Image + Text** use cases.  
+
+### **Setting Up the Environment**
```
Comment (qgallouedec) [inline docs/source/training_vlm_sft.md:55]:
```suggestion
### Setting Up the Environment
```

File: docs/source/training_vlm_sft.md [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+
+In this section, we build the script needed to fine-tune a multimodal model for both **Single Image + Text** and **Multi-Image + Text** use cases.  
+
+### **Setting Up the Environment**  
+
+Before fine-tuning, we need to install the required dependencies. Let's start by setting up the environment:  
+
+```bash
+# Install the required libraries
+pip install -U -q trl transformers datasets bitsandbytes peft accelerate hf_xet
```
Comment (qgallouedec) [inline docs/source/training_vlm_sft.md:61]:
```suggestion
pip install -U -q trl bitsandbytes peft hf_xet
```


these are trl dependencies

File: docs/source/training_vlm_sft.md [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+
+# Alternative: Install a specific Transformers release with zero-day model support  
+# pip install git+https://github.com/huggingface/transformers@v4.49.0-Gemma-3  
+
+# Optional: Install the latest development version of TRL  
+# pip install git+https://github.com/huggingface/trl.git
```
Comment (qgallouedec) [inline docs/source/training_vlm_sft.md:67]: To simplify, you can refer to the installation section of the doc

File: docs/source/training_vlm_sft.md [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+    model_id, 
+    device_map="auto", 
+    torch_dtype=torch.bfloat16, 
+    attn_implementation="eager", 
+    quantization_config=bnb_config
+)
+processor = AutoProcessor.from_pretrained(model_id)
+```
+
+Next, we set up [Quantized Low-Rank Adaptation (QLoRA)](https://arxiv.org/abs/2305.14314), an efficient fine-tuning technique for Large Language Models (LLMs) and Vision-Language Models (VLMs).
```
Comment (qgallouedec) [inline docs/source/training_vlm_sft.md:228]:
```suggestion
Next, we set up [Quantized Low-Rank Adaptation (QLoRA)](https://huggingface.co/papers/2305.14314), an efficient fine-tuning technique for Large Language Models (LLMs) and Vision-Language Models (VLMs).  
```

File: docs/source/training_vlm_sft.md [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+    target_modules="all-linear",
+    task_type="CAUSAL_LM",
+    modules_to_save=[
+        "lm_head",
+        "embed_tokens",
+    ],
+)
+```
+
+With QLoRA now set up, we need to define the training arguments for SFT. The `SFTConfig` class simplifies this process, providing an easy way to adjust parameters based on our specific needs.
```
Comment (qgallouedec) [inline docs/source/training_vlm_sft.md:249]:
```suggestion
With QLoRA now set up, we need to define the training arguments for SFT. The [`SFTConfig`] class simplifies this process, providing an easy way to adjust parameters based on our specific needs.  
```

File: docs/source/training_vlm_sft.md [hunk 10]
```diff
@@ -0,0 +1,386 @@
...
+With QLoRA now set up, we need to define the training arguments for SFT. The `SFTConfig` class simplifies this process, providing an easy way to adjust parameters based on our specific needs.  
+
+```python
+from trl import SFTConfig
+
+training_args = SFTConfig(
+    output_dir="gemma-3-4b-it-trl-sft-llava-instruct-mix-vsft",     # Directory to save the model and push to the Hub. Use a specific repository id (e.g., gemma-3-4b-it-trl-sft-MMIU-Benchmark for multi-image datasets).
+    num_train_epochs=1,                                             # Set the number of epochs to train the model.
+    per_device_train_batch_size=1,                                  # Batch size for each device (e.g., GPU) during training.
+    gradient_accumulation_steps=1,                                  # Number of steps before performing a backward/update pass to accumulate gradients.
```
Comment (qgallouedec) [inline docs/source/training_vlm_sft.md:258]: is it enough, an effective batch size of 1? (assuming you use 1 GPU)

File: docs/source/training_vlm_sft.md [hunk 11]
```diff
@@ -0,0 +1,386 @@
...
+    gradient_checkpointing=True,                                    # Enable gradient checkpointing to reduce memory usage during training.
+    optim="adamw_torch_fused",                                      # Use the fused AdamW optimizer for better performance.
+    logging_steps=10,                                               # Frequency of logging training progress (log every 10 steps).
+    save_strategy="epoch",                                          # Save checkpoints at the end of each epoch.
+    learning_rate=2e-05,                                            # Learning rate for training.
+    bf16=True,                                                      # Enable bfloat16 precision for training to save memory and speed up computations.
+    push_to_hub=True,                                               # Automatically push the fine-tuned model to Hugging Face Hub after training.
+    gradient_checkpointing_kwargs={"use_reentrant": False},         # Set gradient checkpointing to non-reentrant to avoid issues.
+    dataset_text_field="",                                          # Placeholder for dataset text field (needed by the collator).
```
Comment (qgallouedec) [inline docs/source/training_vlm_sft.md:267]:
```suggestion
```

not needed anymore

File: docs/source/training_vlm_sft.md [hunk 12]
```diff
@@ -0,0 +1,386 @@
...
+    dataset_kwargs={"skip_prepare_dataset": True},                  # Skip dataset preparation to handle preprocessing manually.
+)
+
+training_args.remove_unused_columns = False                         # Ensure unused columns are not removed in the collator (important for batch processing).
+
```
Comment (qgallouedec) [inline docs/source/training_vlm_sft.md:272]:
```suggestion
    remove_unused_columns=False,                         # Ensure unused columns are not removed in the collator (important for batch processing).
)
```

File: docs/source/training_vlm_sft.md [hunk 13]
```diff
@@ -0,0 +1,386 @@
...
+    )  # Encode texts and images into tensors
+
+    # The labels are the input_ids, and we mask the padding tokens in the loss computation
+    labels = batch["input_ids"].clone()  # Clone input IDs for labels
+    # Mask image tokens
+    image_token_id = [
+        processor.tokenizer.convert_tokens_to_ids(processor.tokenizer.special_tokens_map["boi_token"])
+    ]
+    # Mask tokens for not being used in the loss computation
+    labels[labels == processor.tokenizer.pad_token_id] = -100
```
Comment (qgallouedec) [inline docs/source/training_vlm_sft.md:334]: Is the eos different from the pad token? because we we to mask the pad tokens, but not the eos tokens.

**Review by qgallouedec — COMMENTED** [review]
Nice! some early reviews

File: docs/source/training_vlm_sft.md (hunk 13 above)
Comment (sergiopaniego) [reply docs/source/training_vlm_sft.md:329]: It is 😄 

Screenshot 2025-04-09 at 15 19 02

**Comment by sergiopaniego** [comment]
Ready for review @qgallouedec 😄 

I've also updated the standalone script since it had some issues when batch_size > 1.

One interesting finding (not sure if it's an issue). 
When using `attn_implementation='eager'`, which is the [recommended](https://github.com/huggingface/transformers/blob/c15a7adb283fa984a40558c7fe7bed30ae975cdd/src/transformers/models/gemma3/modeling_gemma3.py#L934) attn for Gemma 3, if the batch_size > 1, the attention_mask looks something like this (batch_size=4):

```
'attention_mask': tensor([[0, 0, 0,  ..., 1, 1, 1],
        [0, 0, 0,  ..., 1, 1, 1],
        [0, 0, 0,  ..., 1, 1, 1],
        [1, 1, 1,  ..., 1, 1, 1]])
```

It produces NaN losses since it produces -inf for all the values [here](https://github.com/huggingface/transformers/blob/2527f71a47b267f2ea7f4afc71a340106dd08809/src/transformers/models/gemma3/modeling_gemma3.py#L341). It's solved using `processor.tokenizer.padding_side = "right"`

**Comment by qgallouedec** [comment]
> Ready for review @qgallouedec 😄
> 
> I've also updated the standalone script since it had some issues when batch_size > 1.
> 
> One interesting finding (not sure if it's an issue). When using `attn_implementation='eager'`, which is the [recommended](https://github.com/huggingface/transformers/blob/c15a7adb283fa984a40558c7fe7bed30ae975cdd/src/transformers/models/gemma3/modeling_gemma3.py#L934) attn for Gemma 3, if the batch_size > 1, the attention_mask looks something like this (batch_size=4):
> 
> ```
> 'attention_mask': tensor([[0, 0, 0,  ..., 1, 1, 1],
>         [0, 0, 0,  ..., 1, 1, 1],
>         [0, 0, 0,  ..., 1, 1, 1],
>         [1, 1, 1,  ..., 1, 1, 1]])
> ```
> 
> It produces NaN losses since it produces -inf for all the values [here](https://github.com/huggingface/transformers/blob/2527f71a47b267f2ea7f4afc71a340106dd08809/src/transformers/models/gemma3/modeling_gemma3.py#L341). It's solved using `processor.tokenizer.padding_side = "right"`

Interesting indeed. Yes you should use `processor.tokenizer.padding_side = "right"`

**Review by qgallouedec — APPROVED** [review]
Looks good!! Thanks, feel free to merge

File: docs/source/training_vlm_sft.md [hunk 14]
```diff
@@ -0,0 +1,381 @@
...
+
+``` python
+# Training
+from trl import SFTTrainer
+
+trainer = SFTTrainer(
+    model=model,
+    args=training_args,
+    data_collator=collate_fn,
+    train_dataset=dataset["train"], # multi-image -> train_dataset=dataset["test"],
```
[the remark this answers is not in the document]
Comment (sergiopaniego) [reply docs/source/training_vlm_sft.md:349]: The dataset used for multi-image is a benchmark and it only comes with test set.
